graceful/service.go

```diff
@@ -76,69 +76,99 @@ func (s *service) ListenAndServeTLS(ctx context.Context, proto string, addr stri
 	return s.listenAndServe(ctx, proto, addr, httpServer)
 }
 
-func (s *service) ListenAndServe(ctx context.Context, proto string, addr string, handler http.Handler) error {
+func (s *Service) ListenAndServe(ctx context.Context, proto string, addr string, handler http.Handler) error {
 	httpServer := &http.Server{
 		Addr:    addr,
 		Handler: handler,
 	}
 	return s.listenAndServe(ctx, proto, addr, httpServer)
 }
 
-func (s *service) listenAndServe(ctx context.Context, proto string, addr string, server *http.Server) error {
+func (s *Service) listenAndServe(ctx context.Context, _ string, addr string, server *http.Server) error {
+	log := logger.Get(ctx)
+
 	if s.pidFile != "" {
 		pid := os.Getpid()
 		err := os.WriteFile(s.pidFile, []byte(fmt.Sprintf("%d\n", pid)), 0600)
 		if err != nil {
-			return errgo.Notef(err, "fail to write PID file")
+			return errors.Wrap(ctx, err, "fail to write PID file")
 		}
 	}
 
-	ld, err := s.graceful.Listen(proto, addr)
+	// Use tableflip to handle graceful restart requests
+	upg, err := tableflip.New(tableflip.Options{
+		UpgradeTimeout: s.reloadWaitDuration,
+		PIDFile:        s.pidFile,
+	})
 	if err != nil {
-		return errgo.Notef(err, "fail to get listener")
+		log.Fatalln(err)
```

Comment:
Why changing from returning an error, to crashing the software?

Reply:
This is an error on my part. Thanks for picking it up.